Expose Redpanda behind a single NLB

EladDolev opened the pull request:
## Cover letter

This PR tries to add Redpanda operator the capability of exposing Redpanda cluster behind a single NLB
It is very opinionated, placed here for initial feedback, and I am more than happy discussing changes
Once I'll get the initial approval I'll add unit/integration tests

Closes:
- [#1543](https://github.com/vectorizedio/redpanda/discussions/1543)
- [#1777](https://github.com/vectorizedio/redpanda/issues/1777)
- [#1199](https://github.com/vectorizedio/redpanda/issues/1199)
## Release notes

Making the default bootstrap NodePort service route traffic to all brokers
Adding the capabilities for:
- Choosing custom base NodePort per external listener
- Adjusting the way advertised external listeners are being generated
- Creating a NodePort service per broker

EladDolev commented on src/go/k8s/pkg/resources/resource.go:
```diff
@@ -74,7 +75,9 @@ func CreateIfNotExists(
 	err := c.Create(ctx, obj)
 	if err != nil && !errors.IsAlreadyExists(err) {
-		return false, fmt.Errorf("unable to create %s resource: %w", obj.GetObjectKind().GroupVersionKind().Kind, err)
+		if obj.GetObjectKind().GroupVersionKind().Kind == "Service" && !errors.IsInvalid(err) {
```
I guess this is not amazing
would love hearing how this can be improved

senior7515 commented:
@EladDolev we'll review

cc: @i-x

EladDolev commented:
replaces https://github.com/vectorizedio/redpanda/pull/1604

raresfirebolt commented:
Who can help with review on this ?

senior7515 commented:
> Who can help with review on this ?

@i-x michal can we find time to review.

RafalKorepta reviewed (changes requested):
Hi @EladDolev, Thank you for the changes. Overall design seems ok to me. We introduced few changes like schema registry which created merge conflict with your PR. Could you rebase your PR on top of current dev? After you rebase could you split your work into different commits e.g.
* API change and generation
* base nodeport implementation
* Hostname implementation
* ordinal nodeports implementation
* e2e tests
* adjust sample crd (`src/go/k8s/config/sample/external-connectivity.yaml`) to showcase the values or even create new example

On top of your change I would like to propose one unique type of setup where omitting `hostname` and setting `subdomain` to `localhost` with ordinal node port and base node port feature the kind cluster could expose multiple Redpanda nodes to the local development. I don't want to combine this feature in this PR, but rather create follow up.

I left few comments inlined.

EladDolev commented:
> Hi @EladDolev, Thank you for the changes. Overall design seems ok to me. We introduced few changes like schema registry which created merge conflict with your PR. Could you rebase your PR on top of current dev? After you rebase could you split your work into different commits e.g.
> 
> * API change and generation
> * base nodeport implementation
> * Hostname implementation
> * ordinal nodeports implementation
> * e2e tests
> * adjust sample crd (`src/go/k8s/config/sample/external-connectivity.yaml`) to showcase the values or even create new example
> 
> On top of your change I would like to propose one unique type of setup where omitting `hostname` and setting `subdomain` to `localhost` with ordinal node port and base node port feature the kind cluster could expose multiple Redpanda nodes to the local development. I don't want to combine this feature in this PR, but rather create follow up.
> 
> I left few comments inlined.

rebased and created with requested commits
also addressed all of the comments
tests are still missing, so would appreciate it if you could point me in the direction of what exactly is needed
the suggested change might be unnecessary as `localhost.localdomain` can be used to achieve this

mohnda commented:
Hi guys, thanks for the great progress made on this. When is this expected to be released?

joejulian commented:
I'm going to go ahead and close this as it's pretty far behind and I'm not sure if it's still relevant. If you want to pick this back up, please sign the CLA, rebase, and address the comments.
